File: java/src/main/java/com/gaston/macbook/simplemvp/presentation/show_apod/presenter/ApodDetailPresenter.java
package com.gaston.macbook.simplemvp.presentation.show_apod.presenter;

import com.gaston.macbook.simplemvp.base.BasePresenter;
import com.gaston.macbook.simplemvp.presentation.show_apod.ApodDetailContract;
import com.gaston.macbook.simplemvp.presentation.show_apod.model.Apod;
import com.gaston.macbook.simplemvp.domain.interactor.ApodDetailsInteractor;
import com.gaston.macbook.simplemvp.data.cache.ImageCacheImpl;

import androidx.annotation.NonNull;

public class ApodDetailPresenter extends BasePresenter<ApodDetailContract.View> implements ApodDetailContract.Presenter {

    private ApodDetailContract.View view;
    private ApodDetailsInteractor apodInteractor;
    private ImageCacheImpl cache;

    public ApodDetailPresenter(@NonNull ApodDetailContract.View view ,ApodDetailsInteractor apodInteractor, ImageCacheImpl cache) {
        this.view = view;
        this.apodInteractor = apodInteractor;
        this.cache = cache;
    }

    @Override
    public void fetchApodData() {
        view.hideApod();
        view.showProgressBar();
        apodInteractor.getApodDataFromRemote(new ApodDetailsInteractor.onDetailsFetched() {
            @Override
            public void onSuccess(Apod apodFetchedData) {
                    if(isViewAttached()){
                        view.hideProgressBar();
                        view.showApod();
                        view.showApodDetails(apodFetchedData);
                        cache.saveHdUrl(apodFetchedData.getHdurl());
                    }
            }

            @Override
            public void onFailure() {
                if(isViewAttached()){
                    view.showDataFetchError();
                    view.hideProgressBar();
                }

            }
        });
    }


}

File: kotlin/src/main/java/com/gaston/macbook/kotlinsimplemvp/presentation/show_apod/presenter/ApodDetailPresenter.kt
package com.gaston.macbook.kotlinsimplemvp.presentation.show_apod.presenter

import com.gaston.macbook.kotlinsimplemvp.base.BasePresenter
import com.gaston.macbook.kotlinsimplemvp.data.cache.ImageCacheImpl
import com.gaston.macbook.kotlinsimplemvp.domain.interactor.ApodDetailsInteractor
import com.gaston.macbook.kotlinsimplemvp.presentation.show_apod.ApodDetailContract
import com.gaston.macbook.kotlinsimplemvp.presentation.show_apod.model.Apod

class ApodDetailPresenter constructor(view: ApodDetailContract.View, apodInteractor: ApodDetailsInteractor, cache: ImageCacheImpl) : BasePresenter(), ApodDetailContract.Presenter {

    var view: ApodDetailContract.View? = null
    var apodInteractor: ApodDetailsInteractor? = null
    var cache: ImageCacheImpl? = null

    init {
        this.view = view
        this.apodInteractor = apodInteractor
        this.cache = cache
    }

    override fun fetchApodData() {
        view?.hideApod()
        view?.showProgressBar()
        apodInteractor?.getApodDataFromRemote(object : ApodDetailsInteractor.onDetailsFetched {
            override fun onSuccess(apodFetchedData: Apod) {
                view?.hideProgressBar()
                view?.showApod()
                view?.showApodDetails(apodFetchedData)
                cache?.saveHdUrl(apodFetchedData.hdurl)
            }

            override fun onFailure() {
                view?.showDataFetchError()
                view?.hideProgressBar()
            }
        })
    }

}